### DISCUSSION TITLE
Running pyfetch from a callback?

### DISCUSSION THREAD

**Discussion by phorward** [pyodide/pyodide#2215] [Q&A]
Hello all,

I would like to use [pyfetch](url) when a button is clicked, but I have no idea how to handle this case.

Here is my complete example application:
```html
<!doctype html>
<html>
<head>
    <meta charset="UTF-8">
    <script src="https://cdn.jsdelivr.net/pyodide/v0.19.1/full/pyodide.js"></script>
    <script>
	async function main() {
            let pyodide = await loadPyodide({indexURL: "https://cdn.jsdelivr.net/pyodide/v0.19.1/full/"});
            pyodide.runPython(`
                import pyodide
                from js import document

                async def do_fetch(e):
                    res = await pyodide.http.pyfetch("/vi/user/view/self", method="GET")
                    if res.ok:
                        print(res.json())

                e = document.createElement("button")
                e.innerHTML = "Click me"
                e.addEventListener("click", pyodide.create_proxy(do_fetch))

                document.body.appendChild(e)
          `);
        }

	main();
    </script>
</head>
<body>
</body>
</html>
```
When I click the button, I'm getting this error in Firefox:
```
sys:1: RuntimeWarning: coroutine 'do_fetch' was never awaited [pyodide.asm.js:14:296228](https://cdn.jsdelivr.net/pyodide/v0.19.1/full/pyodide.asm.js)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback [pyodide.asm.js:14:296228](https://cdn.jsdelivr.net/pyodide/v0.19.1/full/pyodide.asm.js)
```

Can anybody help?

> With Pyodide 0.18 I've used `js.fetch` for this use case, but this is also not working since Pyodide 0.19 anymore as a Future-object is awaited, resulting in the same problem.

**Comment by hoodmane** [comment]
Thanks for the report @phorward. For now you can do the following:
```py
from js import document
import pyodide
from asyncio import ensure_future
async def do_fetch(e):
    print("hi")
def wrap(e):
    ensure_future(do_fetch(e))
document.body.addEventListener("click", pyodide.create_proxy(wrap))
```
I think there are arguably some bugs here, I will have to look into it.

> With Pyodide 0.18 I've used `js.fetch` for this use case, but this is also not working since Pyodide 0.19 anymore as a Future-object is awaited, resulting in the same problem.

I don't understand what you mean by this could you give example code that worked in v0.18?

**Reply by phorward** [reply]
> For now you can do the following...

Thanks for pointing this out. This one does work as expected:
```python
import pyodide, asyncio
from js import document

async def do_fetch(then):
    res = await pyodide.http.pyfetch("/vi/user/view/self", method="GET")
    if res.ok:
        return then(await res.json())

    then(None)

def do_fetch_wrap(e):
    asyncio.ensure_future(do_fetch(print))

e = document.createElement("button")
e.innerHTML = "Click me"
e.addEventListener("click", pyodide.create_proxy(do_fetch_wrap))

document.body.appendChild(e)
```


> > With Pyodide 0.18 I've used `js.fetch` for this use case, but this is also not working since Pyodide 0.19 anymore as a Future-object is awaited, resulting in the same problem.
> 
> I don't understand what you mean by this could you give example code that worked in v0.18?

Yes, below is a complete example with `js.fetch` using Pyodide 0.18.1. There where no `create_proxy()` or await/async usages required, therefore the code is much shorter with the same result.

```html
<!doctype html>
<html>
<head>
    <meta charset="UTF-8">
    <script src="https://cdn.jsdelivr.net/pyodide/v0.18.1/full/pyodide.js"></script>
    <script>
        async function main() {
            let pyodide = await loadPyodide({indexURL: "https://cdn.jsdelivr.net/pyodide/v0.18.1/full/"});
            pyodide.runPython(`
                from js import document, fetch

                def do_fetch(e):
                    fetch("/vi/user/view/self", method="GET").then(lambda res: res.text().then(lambda res: print(res)))

                e = document.createElement("button")
                e.innerHTML = "Click me"
                e.addEventListener("click", do_fetch)

                document.body.appendChild(e)
          `);
        }

	main();
    </script>
</head>
<body>
</body>
</html>
```

Anyway, your suggestion for solving the current problem was fine, therefore I think this answered my question.

**Comment by hoodmane** [comment]
> `fetch("/vi/user/view/self", method="GET").then(lambda res: res.text().then(lambda res: print(res)))`

Oh that's interesting. So the disadvantage before was that a lot of Python asyncio interfaces expect a `Future` and don't work with a `Promise`. There was also some resource cleanup problem where if the result of a `Promise` is a Python object, the Python object will get leaked because the Promise is not part of the Python garbage collector. So instead we make a `Future` own the result and then things get garbage collected properly.

But we could probably make a special type of `Future` that has `.then` and `.catch` and `.finally` arguments. I will have to think about how that could look. It might be nice to allow people to use either interface...

Also, have you tried:
```py
fetch("/vi/user/view/self", method="GET").add_done_callback(lambda res: res.result().text().add_done_callback(lambda res: print(res.result())))
```
this is closer to your original code, though still longer because of `add_done_callback` and `result`...
